lms/djangoapps/courseware/models.py

```diff
@@ -260,3 +260,20 @@ class Meta:
 
     def __unicode__(self):
         return "[OCGLog] %s: %s" % (self.course_id.to_deprecated_string(), self.created)  # pylint: disable=no-member
+
+
+class StudentFieldOverride(models.Model):
+    """
+    Holds the value of a specific field overriden for a student.  This is used
+    by the code in the `courseware.student_field_overrides` module to provide
+    overrides of xblock fields on a per user basis.
+    """
+    course_id = CourseKeyField(max_length=255, db_index=True)
+    location = LocationKeyField(max_length=255, db_index=True)
+    student = models.ForeignKey(User, db_index=True)
+
+    class Meta:   # pylint: disable=missing-docstring
+        unique_together = (('course_id', 'location', 'student'),)
+
+    field = models.CharField(max_length=255)
+    value = models.TextField(default='null')
```

Comment:
1) The unique constraint creates an ordered, composite index on `(course_id, location, student)`. As modeled, if there are many rows in the database and I do a query for all of a student's overrides in a given course, it will use the individual index on `student` and possibly do an intersection with the `course_id` index -- though I'm not sure how that would play out as the table gets large, given that the `course_id` will have a much lower cardinality. If the composite index were ordered like `(student, course_id, location)`, then a query for all of a particular student's course overrides would just hit the one composite index.

It's not a big issue with the size of data involved with IDDE, but it could become important if this facility were more commonly used.

3) Have you talked to @mulby, @shnayder, or someone on the edX analytics team? We'd need to make sure that these tables are added to the list of data that's exported to researchers. We'll also want to instrument this with events when fields are overridden, so that we can reconstruct historical student state.

4. I agree that adding `field` to the unique index is not useful from a performance perspective. I was thinking more from a data integrity perspective. We can get into a situation where slow request processing and multiple browser clicks can cause multiple gunicorn workers to attempt to add the same row twice (this happens a lot on `StudentModule`). That would cause it to permanently break for that user because every subsequent attempt to call `.get()/.get_or_create()` on it would fail with a `MultipleObjectsReturned` error. If the unique constraint is extended to include `field`, then one of create attempts would throw an `IntegrityError`, and subsequent `.get()` attempts would be fine. 

5. Probably not a requirement, as long as we have events? I defer to @mulby on this.